### DISCUSSION TITLE
WKT support

### DISCUSSION THREAD

**Discussion by iamricard** [thesayyn/protoc-gen-ts#129] [Ideas]
Hi!

I see `Interopability with well knowns.` is on the roadmap. Would you be open to a (hopefully simple) PR that maps WKT imports to `google-protobuf`? For context, this is what we get now:

```ts
// got
import * as dependency_1 from "./../../../../google/protobuf/empty";

// want
import * as emptypb from 'google-protobuf/empty_pb';
```

**Comment by iamricard** [comment]
Actually, I see on the README that this library does support `google-protobuf`. Is it possible I'm doing something wrong?

**Reply by thesayyn** [reply]
Nope, you are doing everything correctly and getting the expected result. that statement in the readme could be clarified further.

**Reply by thesayyn** [reply]
`protoc-gen-ts` uses `google-protobuf` runtime for serialization and deserialization unlike some other generators out there.

**Reply by iamricard** [reply]
Gotcha. So messing around a little bit with it, I see that this won't be a matter of "just" importing from `google-protobuf/google/protobuf/foo_pb`, since the WKT on npm aren't compatible with the types that this library compiles. Is that assessment correct?

If that's the case, is the plan to check in WKT in this repo and compile them here?

**Reply by thesayyn** [reply]
as far as I remember, well-known types had some weird runtime constraints that need to be worked around by `protoc-gen-ts` because they have different interfaces. 
for instance, the generated files that are shipped with `google-protobuf` exports functions to access properties whereas `protoc-gen-ts` use getters. this somehow generates incompatibilities between two.

**Reply by thesayyn** [reply]
yeap that one thing could be done. or generate a shim that interops between two.

**Reply by iamricard** [reply]
For context, it was fairly trivial to get this to happen (the generator code was simple enough to follow!):

```ts
import * as empty_pb from "google-protobuf/google/protobuf/empty_pb";
import * as struct_pb from "google-protobuf/google/protobuf/struct_pb";
import * as pb_1 from "google-protobuf";
export class MessageWithWellKnownTypes extends pb_1.Message {
    constructor(data?: any[] | {
        empty?: empty_pb.Empty;
        struct?: struct_pb.Struct;
    }) {
```

But I soon found out that (de)serialization is just different so it was never going to work.

**Reply by iamricard** [reply]
> yeap that one thing could be done. or generate a shim that interops between two.

Do you have a preference on the approach? We really would prefer to use this library, but WKT compat is a blocker so I'd like to see if I can contribute a patch.

**Reply by thesayyn** [reply]
I can't remember the exact reason why this was failing but I do remember that there was weird runtime code that was generated especially for the WKT.

**Reply by thesayyn** [reply]
you are welcome to find out why this is not working and send a patch. I'd be more than happy to provide assistance, review and merge it with proper testing.

**Reply by thesayyn** [reply]
I agree that this is a blocker for us to be drop-in-replacement for the built-in generator.

**Reply by thesayyn** [reply]
a shim that interops between two is my preferred way.

**Comment by thesayyn** [comment]
This is fixed with the rust rewrite. Now we support well known types!
